File: oss/clawpipe/server/src/api/middleware/security_headers.py
"""Security headers middleware for all Python HTTP services."""

import aiohttp.web

SECURITY_HEADERS = {
    "X-Content-Type-Options": "nosniff",
    "X-Frame-Options": "SAMEORIGIN",
    "Referrer-Policy": "strict-origin-when-cross-origin",
    "Permissions-Policy": "camera=(), microphone=(), geolocation=()",
    "X-XSS-Protection": "0",
    "Cache-Control": "no-store",
    "Content-Security-Policy": (
        "default-src 'none'; "
        "script-src 'self' 'unsafe-inline' cdn.redoc.ly; "
        "style-src 'self' 'unsafe-inline'; "
        "img-src 'self' data:; "
        "font-src 'self'; "
        "connect-src 'self'; "
        "frame-ancestors 'none'"
    ),
    "Strict-Transport-Security": "max-age=63072000; includeSubDomains; preload",
    "Cross-Origin-Resource-Policy": "same-origin",
    "Cross-Origin-Opener-Policy": "same-origin",
}


def security_headers_middleware_factory():
    """Return middleware that adds security headers to all responses."""

    @aiohttp.web.middleware
    async def security_headers_middleware(request, handler):
        response = await handler(request)
        for header, value in SECURITY_HEADERS.items():
            response.headers.setdefault(header, value)
        return response

    return security_headers_middleware

File: products/lunaos/lunaos-engine/packages/api/src/middleware/security-headers.ts
/**
 * Security Headers Middleware — Helmet-equivalent for Cloudflare Workers
 *
 * Adds comprehensive security headers to all responses:
 * - HSTS, CSP, X-Content-Type-Options, X-Frame-Options, etc.
 */

import { createMiddleware } from 'hono/factory';
import type { Env } from '../worker';

export const securityHeaders = createMiddleware<{ Bindings: Env }>(async (c, next) => {
    await next();

    // Strict Transport Security — force HTTPS for 1 year + subdomains
    c.header('Strict-Transport-Security', 'max-age=31536000; includeSubDomains; preload');

    // Content Security Policy — restrict resource loading
    c.header('Content-Security-Policy', [
        "default-src 'none'",
        "script-src 'self'",
        "style-src 'self' 'unsafe-inline'",
        "img-src 'self' data: https:",
        "font-src 'self'",
        "connect-src 'self' https://api.lunaos.ai https://api.lemonsqueezy.com",
        "frame-ancestors 'none'",
        "base-uri 'self'",
        "form-action 'self'",
    ].join('; '));

    // Prevent MIME type sniffing
    c.header('X-Content-Type-Options', 'nosniff');

    // Prevent clickjacking
    c.header('X-Frame-Options', 'DENY');

    // XSS Protection (legacy browsers)
    c.header('X-XSS-Protection', '1; mode=block');

    // Don't send referrer on cross-origin requests
    c.header('Referrer-Policy', 'strict-origin-when-cross-origin');

    // Restrict browser features
    c.header('Permissions-Policy', [
        'camera=()',
        'microphone=()',
        'geolocation=()',
        'payment=(self)',
        'usb=()',
        'magnetometer=()',
        'gyroscope=()',
        'accelerometer=()',
    ].join(', '));

    // Prevent cross-origin attacks
    c.header('Cross-Origin-Opener-Policy', 'same-origin');
    c.header('Cross-Origin-Embedder-Policy', 'require-corp');
    c.header('Cross-Origin-Resource-Policy', 'same-origin');

    // Remove server identification
    c.res.headers.delete('Server');
    c.res.headers.delete('X-Powered-By');
});
